### ISSUE TITLE
Long generated paths break Visual Studio UI

### ISSUE THREAD

**Issue by shuebner** [corvus-dotnet/Corvus.JsonSchema#390]
TL;DR: Corvus could shorten generated file paths for better compatibility with Visual Studio UI.

This is not strictly your fault, but still an issue when working with Corvus-generated code in Visual Studio:

Corvus code generation uses very long file paths.
That can become a problem even when having long paths activated in Windows (and GIT).

I got into the situation today where Visual Studio 2022 (17.10.4) shows red underlined errors in its editor (on the `AsInt32()` method in this case, claiming it does not exist).
It would build successfully though.
Also the dotnet CLI built successfully.

The file where the method was, was a `*.Number.cs` whose full path is 261 characters long.
I think this is not a coincidence.

When I copied the whole code folder to plain `C:\` and the issue was gone.
This makes me believe there is some issue with long file paths (or file names) in Visual Studio itself (maybe related to https://learn.microsoft.com/en-us/answers/questions/1666150/ms-visual-studio-2022-not-able-to-open-source-file).

**Comment by mwadams** [comment]
This is absolutely not a coincidence!

Thanks for opening this issue - I will add something to limit and uniqueify overlong file paths.

It is depressing that we still have to deal with this in 2024...

**Comment by shuebner** [comment]
btw, the long file names also break the VS-internal git diff viewer.

**Comment by mwadams** [comment]
Fixed in #393 - please try the Preview packages on nuget.

**Comment by SJones-CTC** [comment]
Hi, I tried the preview package on nuget using:
dotnet tool install --global Corvus.Json.JsonSchema.TypeGeneratorTool --version 4.0.0-preview.34
Then ran the generatejsonschematypes with the --schemafile --rootNamespace and --rootPath options and it still gave me very long filenames (eg.186 chars). Is there another option I need to add to the command that would result in shorter file names?

**Comment by mwadams** [comment]
(The `PathTruncator` defaults to a 260-character maximum total path length, and truncates. Can you provide an example of a file path it creates that is >260 characters?

**Comment by shaneqld** [comment]
Seeing a similar/related issue in VS 2022 (17.12.14), in a brand new empty ASP.NET template project (default csproj, nothing interesting).

```
The project '...' ran into a problem during the last operation: Cannot expand metadata in expression "$([MSBuild]::ValueOrDefault('%(FullPath)', '').StartsWith($([MSBuild]::EnsureTrailingSlash($(MSBuildProjectDirectory)))))". The item metadata "%(FullPath)" cannot be applied to the path "OpenOutboundNormalizedMessage.ListOfContentElements.OpenOutboundMessageContent.ContentNotificationTemplate_._cationTemplateHeader.TemplateParametersForPlaceholdersInTemplate.NotificationTemplateParameter.ParameterTextValue.Object.cs".

Path:

C:\xxx\xxxxxxxx\xxx\genesys\OpenOutboundNormalizedMessage.ListOfContentElements.OpenOutboundMessageContent.ContentNotificationTemplate_._cationTemplateHeader.TemplateParametersForPlaceholdersInTemplate.NotificationTemplateParameter.ParameterTextValue.Object.cs

exceeds the OS max path limit. The fully qualified file name must be less than 260 characters.

C:\Program Files\dotnet\sdk\9.0.102\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.DefaultItems.targets You may need to reload the solution after fixing the problem. 
```

The full path is exactly 260 characters. I'm not sure if 260 characters includes a null terminating character in some scenarios. If I delete one character from the base directory, everything loads fine.

Source schema is [here](https://developer.genesys.cloud/commdigital/digital/openmessaging/normalizedmsgformat).

**Comment by mwadams** [comment]
I have modified the path truncation to use 259 characters instead of 260, which will hopefully address this problem for you.

The build will be available later today as V4.2.1.
